src/plugins/data/common/index_patterns/types.ts

```diff
@@ -51,3 +52,46 @@ export interface IndexPatternAttributes {
 
 export type OnNotification = (toastInputFields: ToastInputFields) => void;
 export type OnError = (error: Error, toastInputFields: ErrorToastOptions) => void;
+
+export interface FieldFormatSpec {
+  id: string;
+  params: IFieldFormatMetaParams;
+}
+
+export type AggregationRestrictions = Record<
+  string,
+  {
+    agg?: string;
+    interval?: number;
+    fixed_interval?: string;
+    calendar_interval?: string;
+    delay?: string;
+    time_zone?: string;
+  }
+>;
+
+export interface TypeMeta {
+  aggs?: Record<string, AggregationRestrictions>;
+  [key: string]: any;
+}
+
+export interface FieldSpec {
+  [key: string]: any;
+  format?: FieldFormatSpec;
+}
+
+export interface IndexPatternSpec {
+  id?: string;
```

Comment:
why is everything optional ?

Reply:
id and version - saved object properties. unlikely to be present unless persisted

title - there's code in the index pattern class that sets the title attribute to the id if its not present. I'm not sure why an index pattern wouldn't have a title. I assume this is for backward compatibility. I'll require this to be present in the spec.

timeFieldName, sourceFilters, typeMeta - all optional functionality

fields - in case the field list hasn't been created yet.